library/lmots.c

```diff
@@ -0,0 +1,686 @@
+    unsigned short checksum;
+    int rc = 0;
+
+    mbedtls_md_init( &hash_ctx );
+    rc = mbedtls_md_setup( &hash_ctx, mbedtls_md_info_from_type( MBEDTLS_MD_SHA256 ), 0 );
+    if( rc )
+    {
+        goto out;
+    }
+    rc = mbedtls_md_starts( &hash_ctx );
+    if ( rc )
+    {
+        goto out;
+    }
+
+    rc = mbedtls_md_update( &hash_ctx, I_key_identifier, MBEDTLS_LMOTS_I_KEY_ID_LEN );
+    if ( rc )
+    {
+        goto out;
+    }
+
+    rc = mbedtls_md_update( &hash_ctx, q_leaf_identifier, MBEDTLS_LMOTS_Q_LEAF_ID_LEN );
+    if ( rc )
+    {
+        goto out;
+    }
+
+    val_to_network_bytes( D_MESG_CONSTANT, D_CONST_LEN, D_MESG_BYTES );
+    rc = mbedtls_md_update( &hash_ctx, D_MESG_BYTES, sizeof( D_MESG_BYTES ) );
+    if ( rc )
+    {
+        goto out;
+    }
+
+    rc = mbedtls_md_update( &hash_ctx, C_random_value, MBEDTLS_LMOTS_C_RANDOM_VALUE_LEN );
+    if ( rc )
+    {
+        goto out;
+    }
+
+    rc = mbedtls_md_update( &hash_ctx, msg, msg_len );
+    if ( rc )
+    {
+        goto out;
+    }
+
+    rc = mbedtls_md_finish( &hash_ctx, out );
+    if ( rc )
+    {
+        goto out;
+    }
+
+    checksum = lmots_checksum_generate( out );
+    val_to_network_bytes( checksum, CHECKSUM_LEN, out + MBEDTLS_LMOTS_N_HASH_LEN );
+
+out:
+    mbedtls_md_free( &hash_ctx );
+
+    return( rc );
+}
+
+static int hash_symbol_array( const unsigned char I_key_identifier[MBEDTLS_LMOTS_I_KEY_ID_LEN],
+                              const unsigned char q_leaf_identifier[MBEDTLS_LMOTS_Q_LEAF_ID_LEN],
+                              const unsigned char x_symbol_array[MBEDTLS_LMOTS_P_SIG_SYMBOL_LEN][32],
+                              const unsigned char hash_idx_min_values[MBEDTLS_LMOTS_P_SIG_SYMBOL_LEN],
+                              const unsigned char hash_idx_max_values[MBEDTLS_LMOTS_P_SIG_SYMBOL_LEN],
+                              unsigned char output[MBEDTLS_LMOTS_P_SIG_SYMBOL_LEN][32] )
+{
+    unsigned char i_symbol_idx;
+    unsigned char j_hash_idx;
+    unsigned char i_symbol_idx_bytes[I_SYMBOL_IDX_LEN];
+    unsigned char j_hash_idx_bytes[1];
+    unsigned short j_hash_idx_min;
+    unsigned short j_hash_idx_max;
+    mbedtls_md_context_t hash_ctx;
+    unsigned char tmp_hash[32];
+    int rc = 0;
+
+    for ( i_symbol_idx = 0; i_symbol_idx < MBEDTLS_LMOTS_P_SIG_SYMBOL_LEN; i_symbol_idx++ )
+    {
+
+        memcpy( tmp_hash, &x_symbol_array[i_symbol_idx], MBEDTLS_LMOTS_N_HASH_LEN );
+
+        j_hash_idx_min = hash_idx_min_values != NULL ? hash_idx_min_values[i_symbol_idx] : 0;
+        j_hash_idx_max = hash_idx_max_values != NULL ? hash_idx_max_values[i_symbol_idx] : SYMBOL_MAX_VAL;
+
+        for ( j_hash_idx = (unsigned char)j_hash_idx_min; j_hash_idx < j_hash_idx_max; j_hash_idx++ )
+        {
+            mbedtls_md_init( &hash_ctx );
+            rc = mbedtls_md_setup( &hash_ctx, mbedtls_md_info_from_type( MBEDTLS_MD_SHA256 ), 0 );
+            if( rc )
+            {
+                mbedtls_md_free( &hash_ctx );
```

Comment:
This feels awkward, having to have this `mbedtls_md_free` for every failure. This is a little more awkward than this construct normally is, because it is in a loop. Some ideas:

1. Have a cleanup at the end of the function, using goto to get to it, and have a local variable that is a flag indicating if the md needs to be freed.
2. Put the inner part of the loop in another function.

It seems 1 would not affect the code generated too much, other than possibly eliminating multiple calls to the free function. 2 might have extra overhead of the function call, but I suspect on MCU-type devices this is going to all be overwhelmed by the hash function itself, and so probably won't matter.

Reply:
Possible 3, the documentation isn't completely clear to me, but is it possible to do `mbedtls_md_init` at the start of the function, and just do `mbedtls_md_starts` at the beginning of each iteration. This would make cleanup faster, and potentially reduce overhead of deallocation and reallocation of internal state.

Reply:
A standard idiom throughout the library is
```
mbedtls_yyy_init( y );
…
ret = mbedtls_xxx();
if( ret != 0 )
    goto exit;
…
exit:
    mbedtls_yyy_free( y );
    return( ret );
```

That is, @d3zd3z's 3, except that our documentation doesn't guarantee that you can call `mbedtls_md_starts()` after `mbedtls_md_finish()` without calling `mbedtls_md_setup()` again. I think it would work but it's not guaranteed or tested.

Reply:
Fixed

Reply:
> our documentation doesn't guarantee that you can call mbedtls_md_starts() after mbedtls_md_finish() without calling mbedtls_md_setup() again.

I'm curious why we split start and setup into separate functions, other than to be able to re-initialize without having to allocate anything additional.